File: leetcode/3sum/3sum.rb
# LeetCode #15 - 3Sum
# https://leetcode.com/problems/3sum/
# By two pointers, repeated for each suffix, with hash-based duplicate removal.

# @param {Integer[]} nums
# @return {Integer[][]}
def three_sum(nums)
  nums.sort!
  triplets = []

  while nums.size > 2
    first = nums.shift
    pairs = two_sum(nums, -first)
    triplets.concat(pairs.each { |pair| pair.unshift(first) })
  end

  triplets.uniq
end

def two_sum(nums, target)
  pairs = []

  left = 0
  right = nums.size - 1

  while left < right
    sum = nums[left] + nums[right]

    pairs << [nums[left], nums[right]] if sum == target
    left += 1 if sum <= target
    right -= 1 if sum >= target
  end

  pairs
end

File: leetcode/3sum/3sum.java
// LeetCode #15 - 3Sum
// https://leetcode.com/problems/3sum/
// By two pointers, repeated for each suffix, avoiding duplicates.

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> triplets = new ArrayList<>();

        for (var i = 0; i < nums.length - 2; ) {
            var first = nums[i];

            for (var pair : twoSum(nums, i + 1, -first))
                triplets.add(List.of(nums[i], pair[0], pair[1]));

            while (i < nums.length - 2 && nums[i] == first) ++i;
        }

        return triplets;
    }

    private static List<int[]> twoSum(int[] nums, int left, int target) {
        List<int[]> pairs = new ArrayList<>();

        for (var right = nums.length - 1; left < right; ) {
            var first = nums[left];
            var second = nums[right];

            if (first + second == target)
                pairs.add(new int[] { first, second });

            if (first + second <= target)
                while (left < right && nums[left] == first) ++left;

            if (first + second >= target)
                while (left < right && nums[right] == second) --right;
        }

        return pairs;
    }
}
